src/graphene.h

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2018 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_GRAPHENE_H
+#define BITCOIN_GRAPHENE_H
+
+#include "bloom.h"
+#include "config.h"
+#include "consensus/validation.h"
+#include "graphene_set.h"
+#include "iblt.h"
+#include "primitives/block.h"
+#include "protocol.h"
+#include "serialize.h"
+#include "stat.h"
+#include "sync.h"
+#include "uint256.h"
+
+#include <atomic>
+#include <vector>
+
+const unsigned char MIN_MEMPOOL_INFO_BYTES = 8;
+
+class CDataStream;
+class CNode;
+
+class CMemPoolInfo
+{
+public:
+    uint64_t nTx;
+
+public:
+    CMemPoolInfo(uint64_t nTx);
+    CMemPoolInfo();
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream &s, Operation ser_action)
+    {
+        READWRITE(nTx);
+    }
+};
+
+class CGrapheneBlock
+{
+public:
+    CBlockHeader header;
+    std::vector<uint256> vTxHashes; // List of all transactions id's in the block
+    std::vector<CTransactionRef> vAdditionalTxs; // vector of transactions receiver probably does not have
+    uint64_t nBlockTxs;
+    CGrapheneSet *pGrapheneSet;
+
+public:
+    CGrapheneBlock(const CBlockRef pblock, uint64_t nReceiverMemPoolTx);
+    CGrapheneBlock() : pGrapheneSet(nullptr) {}
+    ~CGrapheneBlock();
+    /**
+     * Handle an incoming Graphene block
+     * Once the block is validated apart from the Merkle root, forward the Xpedited block with a hop count of nHops.
+     * @param[in]  vRecv        The raw binary message
+     * @param[in]  pFrom        The node the message was from
+     * @param[in]  strCommand   The message kind
+     * @param[in]  nHops        On the wire, nHops is zero for an incoming Graphene block
+     * @return True if handling succeeded
+     */
+    static bool HandleMessage(CDataStream &vRecv, CNode *pfrom, std::string strCommand, unsigned nHops);
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream &s, Operation ser_action)
+    {
+        READWRITE(header);
+        READWRITE(vAdditionalTxs);
+        READWRITE(nBlockTxs);
```

Comment:
Limit nBlockTxs here (throw an exception if it is out of range).  Otherwise the reserve() in CGrapheneBlock::process could consume lots of memory or raise an out-of-memory exception.

I'm open to suggestions for a reasonable limit, but I think we could assume a min tx size of 100 bytes (can we go bigger), and then use the "excessive block" size to calculate the max number of transactions.